Avoid a validation failure while creating a Wasm array with array.new_elem

ast-hugger opened the pull request:
#### b5148db1c4c10cfd8cdeb2c33c0b26fc1b4223d3
<pre>
Avoid a validation failure while creating a Wasm array with array.new_elem
<a href="https://bugs.webkit.org/show_bug.cgi?id=298930">https://bugs.webkit.org/show_bug.cgi?id=298930</a>
<a href="https://rdar.apple.com/160674746">rdar://160674746</a>

Reviewed by Yusuke Suzuki.

&apos;arrayNewElem&apos; creates the array in two stages: first it creates an array full of null
values, then these transient nulls are replaced with the final values. The problem is that
in the second stage the array is already exposed to GC. If the &quot;official&quot; element type of
the array is not nullable and assertions are enabled, GC validates array contents and the
transient null values cause a failure.

The patch fixes this by adding a flag to JSWebAssemblyArray (if assertions are enabled).
The flag is set by &apos;arrayNewElem&apos; until the array is fully populated with the final
values. Validation code skips an array if it has the flag set.

Even when enabled, the flag does not increase the size of JSWebAssemblyArray. That class
currently has 4 bytes of padding at the end of the object (implicit on 64-bit systems and
explicit on 32-bit). When the flag is enabled, it takes one of those unused bytes.

The patch also adds a GC-aware WasmModuleBuilder required by the tests.

Canonical link: <a href="https://commits.webkit.org/300181@main">https://commits.webkit.org/300181@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/423769e4b8ff8ab05db9d4a62da1325d3b66966e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/121581 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/41278 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/31937 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/128124 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/73666 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/22380c0a-fa71-4a6b-8cce-b57bb4d1e16a) 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/41980 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/49857 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/92406 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/61439 "Passed tests") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/f2309d62-10c2-4d56-baff-939089fbc50c) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/124533 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/33514 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/108897 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/73068 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/c7d81c59-0c53-40a1-bf50-6aeb1638654d) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/32532 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/27069 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/71614 "Built successfully") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/113712 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/103023 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/27236 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/130917 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/120101 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/48509 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/36919 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/100977 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/48877 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/105119 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/100866 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25577 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/46249 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/24331 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/45210 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/48368 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/54080 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/150263 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/47839 "Built successfully") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/25/builds/38355 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/51186 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/49521 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/9e151a860623dbcb608fa4d041221cdf5b467394. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/50893

Constellation commented on Source/JavaScriptCore/wasm/WasmOperationsInlines.h:
```diff
@@ -270,7 +270,13 @@ inline EncodedJSValue arrayNewElem(JSWebAssemblyInstance* instance, uint32_t typ
     if (!array) [[unlikely]]
         return JSValue::encode(jsNull());
+#if ASSERT_ENABLED
+    array->setIsUnpopulated(true);
```
You need to put `WTF::storeStoreFence()` after this to ensure that this `true` is published already for the concurrent collectors when executing `copyElementSegment` code.

ast-hugger replied:
Great, I guess it makes sense for it to go right into `setIsUnpopulated()`.

ast-hugger replied:
Added a `storeLoadFence()`, as we need a stronger ordering here than `storeStore` because any future load of the flag needs to happen after both the stores of nulls into the array and the store of `true` into the flag.

Constellation commented on Source/JavaScriptCore/wasm/WasmOperationsInlines.h:
```diff
@@ -270,7 +270,13 @@ inline EncodedJSValue arrayNewElem(JSWebAssemblyInstance* instance, uint32_t typ
+#endif
     instance->copyElementSegment(array, instance->module().moduleInformation().elements[elemSegmentIndex], offset, arraySize, array->span<uint64_t>().data());
+#if ASSERT_ENABLED
+    array->setIsUnpopulated(false);
```
And probably ditto before this.

Constellation reviewed (approved):
r=me

ast-hugger committed:
Avoid a validation failure while creating a Wasm array with array.new_elem
https://bugs.webkit.org/show_bug.cgi?id=298930
rdar://160674746

Reviewed by Yusuke Suzuki.

'arrayNewElem' creates the array in two stages: first it creates an array full of null
values, then these transient nulls are replaced with the final values. The problem is that
in the second stage the array is already exposed to GC. If the "official" element type of
the array is not nullable and assertions are enabled, GC validates array contents and the
transient null values cause a failure.

The patch fixes this by adding a flag to JSWebAssemblyArray (if assertions are enabled).
The flag is set by 'arrayNewElem' until the array is fully populated with the final
values. Validation code skips an array if it has the flag set.

Even when enabled, the flag does not increase the size of JSWebAssemblyArray. That class
currently has 4 bytes of padding at the end of the object (implicit on 64-bit systems and
explicit on 32-bit). When the flag is enabled, it takes one of those unused bytes.

The patch also adds a GC-aware WasmModuleBuilder required by the tests.

Canonical link: https://commits.webkit.org/300181@main
